app/common.py

```diff
@@ -0,0 +1,318 @@
+    2. it deals with the same symlinks by checking the link target,
+        re-generate the symlink if the dst symlink is not the same
+        as the src.
+    3. it will remove files that not presented in the src, and
+        unconditionally override files with same path, ensuring
+        that the dst will be identical with the src.
+
+    NOTE: is_file/is_dir also returns True if it is a symlink and
+    the link target is_file/is_dir
+    """
+    if dst.is_symlink() or not dst.is_dir():
+        raise FileNotFoundError(f"{dst} is not found or not a dir")
+
+    # phase1: populate files to the dst
+    for cur_dir, _, files in os.walk(src, topdown=True, followlinks=False):
+        _cur_dir = Path(cur_dir)
+        _cur_dir_on_dst = dst / _cur_dir.relative_to(src)
+
+        # cover the edge case that dst is not a dir.
+        if _cur_dir_on_dst.is_symlink() or not _cur_dir_on_dst.is_dir():
+            _cur_dir_on_dst.unlink(missing_ok=True)
+            _cur_dir_on_dst.mkdir(parents=True)
+            copy_stat(_cur_dir, _cur_dir_on_dst)
+
+        # populate files
+        for fname in files:
+            _src_f = _cur_dir / fname
+            _dst_f = _cur_dir_on_dst / fname
+
+            # prepare dst
+            #   src is file but dst is a folder
+            #   delete the dst in advance
+            if not _dst_f.is_symlink() and _dst_f.is_dir():
+                shutil.rmtree(_dst_f, ignore_errors=True)
+
+            # copy/symlink dst as src
+            #   if src is symlink, check symlink, re-link if needed
+            if _src_f.is_symlink():
+                dst_symlink_as_src(_src_f, _dst_f)
+            else:
+                # copy/override src to dst
+                shutil.copy(_src_f, _dst_f, follow_symlinks=False)
+                copy_stat(_src_f, _dst_f)
+
+    # phase2: remove unused files in the dst
+    for cur_dir, dirs, files in os.walk(dst, topdown=True, followlinks=False):
+        _cur_dir_on_dst = Path(cur_dir)
+        _cur_dir_on_src = src / _cur_dir_on_dst.relative_to(dst)
+
+        # remove unused dir
+        if not _cur_dir_on_src.is_dir():
+            shutil.rmtree(_cur_dir_on_dst, ignore_errors=True)
+            dirs.clear()  # stop iterate the subfolders of this dir
+            continue
+
+        for fname in files:
+            _src_f = _cur_dir_on_src / fname
+            if not (_src_f.is_symlink() or _src_f.is_file()):
+                (_cur_dir_on_dst / fname).unlink(missing_ok=True)
+
+
+def re_symlink_atomic(link: Path, target: Union[Path, str]):
+    """Re-link a symlink to new destination atomically.
+
+    NOTE: shutil.move use os.rename, and os.rename is atomic when
+    src and dst are on the same filesystem under linux.
+    """
+    if not (link.is_symlink() and link.readlink() == Path(target)):
+        tmp_link = Path(link).parent / f"tmp_link_{os.urandom(6)}"
+        try:
+            tmp_link.symlink_to(target)
+            shutil.move(tmp_link, link)
```

Comment:
It seems that this function is not used. I'm afraid that if the argument `link` points to a directory, this shutil doesn't work as expected.
`tmp_link` is moved under the directory to which the `link` is pointed.

Reply:
Change to use `os.rename` instead of `shutil.move`, so the scenario of symlink to `dir` should be covered.